### PR TITLE
Add max and min values to Statistics

### PR DESCRIPTION
**Pull request by viirya**
# Description

This patch adds max and min values to Statistics.

### CODE REVIEW

**Comment by houqp** [comment]
Recommend installing a rust-fmt plugin for the editor you use so you won't need to manually fix formatting ever again :)

**Comment by houqp** [comment]
also cc @Dandandan

File: rust/src/delta_datafusion.rs [hunk 1]
```diff
@@ -163,9 +213,51 @@ impl TableProvider for delta::DeltaTable {
...
 }
 
-fn stat_to_val(stat: &ColumnCountStat) -> Option<usize> {
-    match stat {
-        ColumnCountStat::Value(val) => Some(*val as usize),
-        ColumnCountStat::Column(_) => None,
+fn to_scalar_value(stat_val: &serde_json::Value) -> Option<datafusion::scalar::ScalarValue> {
+    if stat_val.is_number() {
+        if stat_val.is_i64() {
+            Some(ScalarValue::from(stat_val.as_i64()?))
```
Comment (houqp) [inline rust/src/delta_datafusion.rs:219]: I think the `is_type()` method call can be avoided if we reuse the option from `as_type` method:

```suggestion
        if let Some(val) = stat_val.as_i64() {
            Some(ScalarValue::from(val))
```

File: rust/src/delta_datafusion.rs [hunk 2]
```diff
@@ -147,12 +153,56 @@ impl TableProvider for delta::DeltaTable {
...
+                                                        ) {
+                                                            Some(old_min_value)
+                                                        } else {
+                                                            Some(min_value)
+                                                        }
+                                                    }
+                                                    _ => Some(min_value),
+                                                },
+                                                _ => old_stats.min_value,
+                                            }
```
Comment (houqp) [inline rust/src/delta_datafusion.rs:202]: nitpick, you could reduce nested level and make the code more readable by matching on two values at the same time:

```suggestion
                                            match (min_value, old_stats.min_value) {
                                                (Some(min_value), Some(old_min_value)) => {
                                                    if left_larger_than_right(
                                                        min_value.clone(),
                                                        old_min_value.clone(),
                                                    ) {
                                                        Some(old_min_value)
                                                    } else {
                                                        Some(min_value)
                                                    }
                                                }
                                                (Some(min_value), None) => Some(min_value),
                                                (None, old) => old,
                                            }
```

File: rust/src/delta_datafusion.rs [hunk 3]
```diff
@@ -163,9 +213,51 @@ impl TableProvider for delta::DeltaTable {
...
+        }
+        ScalarValue::Int32(Some(v)) => {
+            let i_right = i32::try_from(right).unwrap();
+            v > i_right
+        }
+        ScalarValue::Int64(Some(v)) => {
+            let i_right = i64::try_from(right).unwrap();
+            v > i_right
+        }
+        _ => unimplemented!("Unimplemented for {:?}", left),
```
Comment (houqp) [inline rust/src/delta_datafusion.rs:261]: nitpick, could we make this error message more explicit for easier debugging?

```suggestion
        _ => unimplemented!("Scalar value comparison unimplemented for {:?} and {:?}", left, right),
```

**Review by houqp — CHANGES_REQUESTED** [review]

**Review by houqp — APPROVED** [review]
LGTM, I will give @Dandandan some time to review before merging this in. Thanks @viirya !

**Comment by viirya** [comment]
Thank you @houqp!

File: rust/src/delta_datafusion.rs [hunk 4]
```diff
@@ -163,9 +209,53 @@ impl TableProvider for delta::DeltaTable {
...
+fn to_scalar_value(stat_val: &serde_json::Value) -> Option<datafusion::scalar::ScalarValue> {
+    if stat_val.is_number() {
+        if let Some(val) = stat_val.as_i64() {
+            Some(ScalarValue::from(val))
+        } else if let Some(val) = stat_val.as_u64() {
+            Some(ScalarValue::from(val))
+        } else {
+            stat_val.as_f64().map(ScalarValue::from)
+        }
+    } else {
```
Comment (Dandandan) [inline rust/src/delta_datafusion.rs:221]: Probably some more easy to add are string / boolean.

Comment (viirya) [reply rust/src/delta_datafusion.rs:221]: Yea, just wondering if max/min values of string/boolean are kept in statistics too.

File: rust/src/delta_datafusion.rs [hunk 5]
```diff
@@ -163,9 +209,53 @@ impl TableProvider for delta::DeltaTable {
     }
 }
 
-fn stat_to_val(stat: &ColumnCountStat) -> Option<usize> {
-    match stat {
-        ColumnCountStat::Value(val) => Some(*val as usize),
-        ColumnCountStat::Column(_) => None,
+fn to_scalar_value(stat_val: &serde_json::Value) -> Option<datafusion::scalar::ScalarValue> {
```
Comment (Dandandan) [inline rust/src/delta_datafusion.rs:212]: I think to be complete we have to pass the (arrow) datatype as well to this function, for example to parse/handle timestamps properly.

Comment (viirya) [reply rust/src/delta_datafusion.rs:212]: Hm, for comparison purpose, I think using few basic primitive types are more easy. Maybe we can convert the compared `ScalarValue` to correct `ScalarValue` after comparison, e.g. `ScalarValue::Int64` to `ScalarValue::TimestampSecond` if it is Timestamp Second, etc.

Comment (viirya) [reply rust/src/delta_datafusion.rs:212]: How about addressing this in another PR?

Comment (Dandandan) [reply rust/src/delta_datafusion.rs:212]: Yeah perfectly fine to handle this in a follow up PR.

FYI, There was some other approach of using the aggregation code from DataFusion over here:
https://github.com/apache/arrow-datafusion/pull/719

Might also make sense to extend the ScalarValue a bit at DataFusion side  (e.g. derive `Ord`)

**Review by Dandandan — APPROVED** [review]

File: rust/src/delta_datafusion.rs (hunk 5 above)
Comment (houqp) [reply rust/src/delta_datafusion.rs:212]: +1 on extending ScalarValue in Datafusion wherever applicable for better code reuse. Implementing `Ord` trait there is a great idea.

Comment (viirya) [reply rust/src/delta_datafusion.rs:212]: Sounds good, the suggestion. I will find some time to see if I can implement `Ord` there.

**Pull request merged**